core/types/transaction.go

```diff
@@ -105,11 +105,11 @@ type TransactionMisc struct {
 	time time.Time // Time first seen locally (spam avoidance)
 
 	// caches
-	hash atomic.Value //nolint:structcheck
-	from atomic.Value
+	hash atomic.Pointer[libcommon.Hash] //nolint:structcheck
+	from atomic.Pointer[libcommon.Address]
 
 	// cache how much gas the tx takes on L1 for its share of rollup data
-	rollupGas atomic.Value
+	rollupGas atomic.Pointer[types2.RollupCostData]
```

Comment:
Thanks for your contribution. However, I spot some few issues and feel like sharing the overall op-erigon maintenance policy would be beneficial for all of us.

Our policy to minimize code diff from the original [erigon](https://github.com/ledgerwatch/erigon) implementation. This helps use to narrow down optimism specific code and helps us to maintain.

In this PR, I see that you have altered three field types: `hash`, `from`, `rollupGas` from `atomic.Value` to `atomic.Pointer[T]`. Your intention was to speed up the `Store` operation, and in that sense you are correct: `atomic.Pointer` store is way cheaper that `atomic.Value` store.

However, this change adds diff comparing with upstream. We will have merge conflicts every time while fetching erigon upstream. Can you please only apply the type changes only to optimism specific fields: `rollupGas`? After that, can you please test that the CPU issue is solved? In our side, this issue was not reproducible.

If only we change the type of `rollupGas` and issue still persists, this is a problem and we shall dive in deeper. If is not, we are all almost happy, but still have some issues.

Failed CI's linter is complaining about `atomic.Pointer` must not be copied. The proper way to do this is to fix `copy` method of transactions, and manually copy the `atomic.Pointer` fields. I believe that the original erigon code has same problem, although for legacy golang issues, linter could not catch this error. I think this copy bug can be first fixed at upstream(only for `hash`, `from` fields since upstream does not know about optimism), and fetch down to op-erigon.

TL; DR:
1. Our policy is to minimize diff with upstream erigon.
2. Lets see the bug is fixed when only updating the type of `rollupGas`; to minimize diff.
3. Still, bug persists because atomic types cannot be copied. This must be also fixed.

Reply:
1. Totally make sense
2. I did a commit with only changes to rollupGasg. I'm currently stress-testing it and will profile again to see if it fixes the issue.
3. From https://pkg.go.dev/sync/atomic regarding atomic.Value: ```Once Store has been called, a Value must not be copied. A Value must not be copied after first use.```
It seems like upstream Erigon does not respect this already from what I see. There should be no change of behaviour with my change (except that it triggers the linter)

Reply:
For point 3, opening a PR to upstream to fix the issue seems to be valid. I think if you want to contribute upstream as well, please go ahead. If not, I shall open the PR when I'm available. Maybe opening an issue would be enough.

Reply:
Also, Can you run CIs again for the new commits? I think `nocopy` issue still persists, because of usage of `atomic.Pointer`. Expecting linter to complain.

Reply:
Oh, I found I can trigger CIs.

Reply:
Sure, here it is: [profile.pb.gz](https://github.com/testinprod-io/op-erigon/files/14461478/profile.pb.gz)

Regarding linter, as you said, it's related to `nocopy`, but the actual behaviour is the same as previous versions of the copy do copy the transactions structs holding `atomic.Value`.